jni/include/native_engines_stream_support.h

```diff
@@ -119,6 +120,105 @@ class NativeEngineIndexInputMediator {
 
 
 
+/**
+ * This class delegates the provided index output to do IO processing.
+ * In most cases, it is expected that IndexOutputWithBuffer was passed down to this,
+ * which eventually have Lucene's IndexOutput to write bytes.
+ */
+class NativeEngineIndexOutputMediator {
+ public:
+  NativeEngineIndexOutputMediator(JNIUtilInterface *_jni_interface,
+                                  JNIEnv *_env,
+                                  jobject _indexOutput)
+      : jni_interface(_jni_interface),
+        env(_env),
+        indexOutput(_indexOutput),
+        bufferArray((jbyteArray) (_jni_interface->GetObjectField(_env,
+                                                                 _indexOutput,
+                                                                 getBufferFieldId(_jni_interface, _env)))),
+        writeBytesMethod(getWriteBytesMethod(_jni_interface, _env)),
+        bufferLength(jni_interface->GetJavaBytesArrayLength(env, bufferArray)),
+        nextWriteIndex() {
+  }
+
+  void writeBytes(uint8_t *source, size_t nbytes) {
+    auto left = nbytes;
+    while (left > 0) {
+      const auto writeBytes = std::min(bufferLength - nextWriteIndex, left);
+
+      // === Critical Section Start ===
+
+      // Get primitive array pointer, no copy is happening in OpenJDK.
+      auto primitiveArray =
```

Comment:
This seems risky to me. From https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html#GetPrimitiveArrayCritical_ReleasePrimitiveArrayCritical:
> If possible, the VM returns a pointer to the primitive array; otherwise, a copy is made.

What happens if this primitiveArray is a copy? OpenSearch can run potentially with JDKs other than OpenJDK...

Should, instead, we try to use `MemorySegment` somehow instead?

Reply:
sure, I understand your concern, but it's very likely that the most JDKs out there will return their internal pointer rather than returning a copied version. 
I think we can trade-off this to serve most frequent cases with just one copy. If we introduce an intermediate buffer like MemorySegment or native ByteBuffer, then we need two copies. bytes -> intermediate -> byte[] for IndexInput.

But I think I missed to pass `0` instead of `JNI_COMMIT` to ensure releasing the acquired pointer after receiving a copied one. (`0` will be ignored otherwise, so we can just pass `0`)
```
mode	actions
0	copy back the content and free the elems buffer
JNI_COMMIT	copy back the content but do not free the elems buffer
JNI_ABORT	free the buffer without copying back the possible changes
```

What do you think?

Solution 1 (Preferred).
Check whether the returned is copied, then pass `jint mode` accordingly.
Pros :
  - Will work most cases. Modern JDK will return a primitive pointer.
Cons:
  - In case JDK not supporting returning a primitive pointer, it will involve several memory allocations.


Solution 2
Have an intermediate buffer then convert it to `byte[]` for IndexOutput.
Pros : 
  - Will cover all the JDK implementations.
Cons :
  - Even when JDK supports a primitive array, we still need two copies.

Reply:
I like solution 1 - to confirm, it will work for all jdks in general right?

Im wondering in future if we could leverage panama MemorySegment so that we can manage the buffer on the java side. Maybe for another review

Reply:
Hi Jack, 
Even with MemorySegment, it will need at least one copy.
Since it is native engine that uses IOWriter to write bytes, the given bytes from it locates at a different region in memory. So we still need to copy those bytes to the region that MemorySegment is pointing to. (Java does not provide a way for us to reset the internal pointer)

After copied bytes into MemorySegment, Java can directly access to via `get() -> byte` method, but given that the index size can grow giga byte scale, I think just copy bytes in bulk would be better than getting single byte then pass it.

Reply:
Also since it is bytes copying, there's no endianess restriction on it, it will be guaranteed that byte[] will be always identical to uint8_t[] what we are given.

Guarantee:
```
len(byte[]) == len(uint8_t[])
and for any i, byte[i] == uint8_t[i]
```